py3/unicode errors

running the sqlalchemy 0.9.9 testsuite with mysqlclient HEAD I got the following errors
(see https://bitbucket.org/zzzeek/sqlalchemy/issue/3358/py3k-support-for-mysqldb-via-mysqlclient#comment-17218449)
```
=================================== FAILURES ===================================
________________ TypesTest_mysql_mysqldb.test_bit_50_roundtrip _________________
Traceback (most recent call last):
  File "<string>", line 2, in test_bit_50_roundtrip
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 84, in decorate
    return self._do(config._current, fn, *args, **kw)
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 114, in _do
    self._expect_failure(config, ex, name=fn.__name__)
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 126, in _expect_failure
    util.raise_from_cause(ex)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 188, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=exc_value)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 182, in reraise
    raise value
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 112, in _do
    return_value = fn(*args, **kw)
  File "<string>", line 2, in test_bit_50_roundtrip
  File "$path/test/../lib/sqlalchemy/testing/util.py", line 191, in provide_metadata
    return fn(*args, **kw)
  File "$path/test/dialect/mysql/test_types.py", line 337, in test_bit_50_roundtrip
    roundtrip([0, 0, 0, 0, i, i, i, i])
  File "$path/test/dialect/mysql/test_types.py", line 321, in roundtrip
    row = table.select().execute().first()
  File "$path/test/../lib/sqlalchemy/sql/base.py", line 386, in execute
    return e._execute_clauseelement(self, multiparams, params)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1870, in _execute_clauseelement
    return connection._execute_clauseelement(elem, multiparams, params)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 938, in _execute_clauseelement
    compiled_sql, distilled_params
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1274, in _handle_dbapi_exception
    util.reraise(*exc_info)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 182, in reraise
    raise value
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 373, in _query
    self._post_get_result()
  File "$site/MySQLdb/cursors.py", line 377, in _post_get_result
    self._rows = self._fetch_row(0)
  File "$site/MySQLdb/cursors.py", line 345, in _fetch_row
    return self._result.fetch_row(size, self._fetch_type)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
__________________ CompiledCacheTest_mysql_mysqldb.test_cache __________________
Traceback (most recent call last):
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
_mysql_exceptions.OperationalError: (1054, "Unknown column 'user_name' in 'field list'")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 730, in test_cache
    cached_conn.execute(ins, {'user_name': 'u1'})
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 841, in execute
    return meth(self, multiparams, params)
  File "$path/test/../lib/sqlalchemy/sql/elements.py", line 322, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 938, in _execute_clauseelement
    compiled_sql, distilled_params
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1271, in _handle_dbapi_exception
    exc_info
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 188, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=exc_value)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 181, in reraise
    raise value.with_traceback(tb)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.OperationalError: (OperationalError) (1054, "Unknown column 'user_name' in 'field list'") b'INSERT INTO users (user_name) VALUES (%s)' ('u1',)
______ CompiledCacheTest_mysql_mysqldb.test_keys_independent_of_ordering _______
Traceback (most recent call last):
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
_mysql_exceptions.OperationalError: (1054, "Unknown column 'user_name' in 'field list'")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 741, in test_keys_independent_of_ordering
    {"user_id": 1, "user_name": "u1", "extra_data": "e1"})
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 841, in execute
    return meth(self, multiparams, params)
  File "$path/test/../lib/sqlalchemy/sql/elements.py", line 322, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 938, in _execute_clauseelement
    compiled_sql, distilled_params
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1271, in _handle_dbapi_exception
    exc_info
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 188, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=exc_value)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 181, in reraise
    raise value.with_traceback(tb)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.OperationalError: (OperationalError) (1054, "Unknown column 'user_name' in 'field list'") b'INSERT INTO users (user_id, user_name, extra_data) VALUES (%s, %s, %s)' (1, 'u1', 'e1')
_____ EngineEventsTest_mysql_mysqldb.test_cursor_events_ctx_execute_scalar _____
Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 1197, in test_cursor_events_ctx_execute_scalar
    [call(conn, ctx.cursor, stmt, ctx.parameters[0], ctx, False)])
  File "$path/test/../lib/sqlalchemy/testing/assertions.py", line 187, in eq_
    assert a == b, msg or "%r != %r" % (a, b)
AssertionError: [call(<sqlalchemy.engine.base.Connection object at 0x1071d96a0>, <MySQLdb.cursors.Cursor object at 0x106bc29b0>, b'SELECT 1', (), <sqlalchemy.dialects.mysql.mysqldb.MySQLExecutionContext_mysqldb object at 0x106c25a90>, False)] != [call(<sqlalchemy.engine.base.Connection object at 0x1071d96a0>, <MySQLdb.cursors.Cursor object at 0x106bc29b0>, 'SELECT 1', (), <sqlalchemy.dialects.mysql.mysqldb.MySQLExecutionContext_mysqldb object at 0x106c25a90>, False)]
__________ EngineEventsTest_mysql_mysqldb.test_cursor_events_execute ___________
Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 1216, in test_cursor_events_execute
    [call(conn, ctx.cursor, stmt, ctx.parameters[0], ctx, False)])
  File "$path/test/../lib/sqlalchemy/testing/assertions.py", line 187, in eq_
    assert a == b, msg or "%r != %r" % (a, b)
AssertionError: [call(<sqlalchemy.engine.base.Connection object at 0x1071fc048>, <MySQLdb.cursors.Cursor object at 0x10719dcc0>, b'SELECT 1', (), <sqlalchemy.dialects.mysql.mysqldb.MySQLExecutionContext_mysqldb object at 0x106c17710>, False)] != [call(<sqlalchemy.engine.base.Connection object at 0x1071fc048>, <MySQLdb.cursors.Cursor object at 0x10719dcc0>, 'SELECT 1', (), <sqlalchemy.dialects.mysql.mysqldb.MySQLExecutionContext_mysqldb object at 0x106c17710>, False)]
______________ EngineEventsTest_mysql_mysqldb.test_execute_events ______________
Traceback (most recent call last):
  File "<string>", line 2, in test_execute_events
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 84, in decorate
    return self._do(config._current, fn, *args, **kw)
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 114, in _do
    self._expect_failure(config, ex, name=fn.__name__)
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 126, in _expect_failure
    util.raise_from_cause(ex)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 188, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=exc_value)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 182, in reraise
    raise value
  File "$path/test/../lib/sqlalchemy/testing/exclusions.py", line 112, in _do
    return_value = fn(*args, **kw)
  File "$path/test/engine/test_execute.py", line 1313, in test_execute_events
    self._assert_stmts(cursor, cursor_stmts)
  File "$path/test/engine/test_execute.py", line 1079, in _assert_stmts
    assert False, "Nothing available for stmt: %s" % stmt
AssertionError: Nothing available for stmt: INSERT INTO t1 (c1, c2)
assert False
__________ ExecuteTest_mysql_mysqldb.test_unicode_test_fails_warning ___________
Traceback (most recent call last):
  File "$path/test/../lib/sqlalchemy/pool.py", line 951, in _do_get
    return self._pool.get(wait, self._timeout)
  File "$path/test/../lib/sqlalchemy/util/queue.py", line 145, in get
    raise Empty
sqlalchemy.util.queue.Empty

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 522, in test_unicode_test_fails_warning
    eng.connect
  File "$path/test/../lib/sqlalchemy/testing/assertions.py", line 224, in assert_raises_message
    callable_(*args, **kwargs)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1890, in connect
    return self._connection_cls(self, **kwargs)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 60, in __init__
    self.__connection = connection or engine.raw_connection()
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1964, in raw_connection
    return self.pool.unique_connection()
  File "$path/test/../lib/sqlalchemy/pool.py", line 280, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "$path/test/../lib/sqlalchemy/pool.py", line 645, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "$path/test/../lib/sqlalchemy/pool.py", line 440, in checkout
    rec = pool._do_get()
  File "$path/test/../lib/sqlalchemy/pool.py", line 964, in _do_get
    return self._create_connection()
  File "$path/test/../lib/sqlalchemy/pool.py", line 285, in _create_connection
    return _ConnectionRecord(self)
  File "$path/test/../lib/sqlalchemy/pool.py", line 416, in __init__
    exec_once(self.connection, self)
  File "$path/test/../lib/sqlalchemy/event/attr.py", line 250, in exec_once
    self(*args, **kw)
  File "$path/test/../lib/sqlalchemy/event/attr.py", line 260, in __call__
    fn(*args, **kw)
  File "$path/test/../lib/sqlalchemy/util/langhelpers.py", line 1219, in go
    return once_fn(*arg, **kw)
  File "$path/test/../lib/sqlalchemy/engine/strategies.py", line 166, in first_connect
    dialect.initialize(c)
  File "$path/test/../lib/sqlalchemy/dialects/mysql/base.py", line 2453, in initialize
    self._detect_ansiquotes(connection)
  File "$path/test/../lib/sqlalchemy/dialects/mysql/base.py", line 2718, in _detect_ansiquotes
    connection.execute("SHOW VARIABLES LIKE 'sql_mode'"),
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 833, in execute
    return self._execute_text(object, multiparams, params)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 982, in _execute_text
    statement, parameters
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1274, in _handle_dbapi_exception
    util.reraise(*exc_info)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 182, in reraise
    raise value
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$path/test/engine/test_execute.py", line 514, in execute
    if "test unicode returns" in stmt:
TypeError: 'str' does not support the buffer interface
______ HandleErrorTest_mysql_mysqldb.test_exception_event_non_dbapi_error ______
Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 1827, in test_exception_event_non_dbapi_error
    eq_(ctx.statement, "select ")
  File "$path/test/../lib/sqlalchemy/testing/assertions.py", line 187, in eq_
    assert a == b, msg or "%r != %r" % (a, b)
AssertionError: b'select ' != 'select '
----------------------------- Captured stdout call -----------------------------
b"I'm not a DBAPI error"
_______________ HandleErrorTest_mysql_mysqldb.test_handle_error ________________
Traceback (most recent call last):
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
_mysql_exceptions.ProgrammingError: (1146, "Table 'test_schema.i_dont_exist' doesn't exist")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 1638, in test_handle_error
    conn.execute("SELECT FOO FROM I_DONT_EXIST")
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 833, in execute
    return self._execute_text(object, multiparams, params)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 982, in _execute_text
    statement, parameters
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1271, in _handle_dbapi_exception
    exc_info
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 188, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=exc_value)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 181, in reraise
    raise value.with_traceback(tb)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.ProgrammingError: (ProgrammingError) (1146, "Table 'test_schema.i_dont_exist' doesn't exist") b'SELECT FOO FROM I_DONT_EXIST' ()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 1645, in test_handle_error
    eq_(ctx.statement, "SELECT FOO FROM I_DONT_EXIST")
  File "$path/test/../lib/sqlalchemy/testing/assertions.py", line 187, in eq_
    assert a == b, msg or "%r != %r" % (a, b)
AssertionError: b'SELECT FOO FROM I_DONT_EXIST' != 'SELECT FOO FROM I_DONT_EXIST'
____________ HandleErrorTest_mysql_mysqldb.test_legacy_dbapi_error _____________
Traceback (most recent call last):
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
_mysql_exceptions.ProgrammingError: (1146, "Table 'test_schema.i_dont_exist' doesn't exist")

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 1576, in test_legacy_dbapi_error
    conn.execute("SELECT FOO FROM I_DONT_EXIST")
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 833, in execute
    return self._execute_text(object, multiparams, params)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 982, in _execute_text
    statement, parameters
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1271, in _handle_dbapi_exception
    exc_info
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 188, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=exc_value)
  File "$path/test/../lib/sqlalchemy/util/compat.py", line 181, in reraise
    raise value.with_traceback(tb)
  File "$path/test/../lib/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "$path/test/../lib/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
  File "$site/MySQLdb/cursors.py", line 220, in execute
    self.errorhandler(self, exc, value)
  File "$site/MySQLdb/connections.py", line 36, in defaulterrorhandler
    raise errorvalue
  File "$site/MySQLdb/cursors.py", line 209, in execute
    r = self._query(query)
  File "$site/MySQLdb/cursors.py", line 372, in _query
    rowcount = self._do_query(q)
  File "$site/MySQLdb/cursors.py", line 335, in _do_query
    db.query(q)
  File "$site/MySQLdb/connections.py", line 280, in query
    _mysql.connection.query(self, query)
sqlalchemy.exc.ProgrammingError: (ProgrammingError) (1146, "Table 'test_schema.i_dont_exist' doesn't exist") b'SELECT FOO FROM I_DONT_EXIST' ()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "$path/test/engine/test_execute.py", line 1580, in test_legacy_dbapi_error
    eq_(canary.mock_calls[0][1][2], "SELECT FOO FROM I_DONT_EXIST")
  File "$path/test/../lib/sqlalchemy/testing/assertions.py", line 187, in eq_
    assert a == b, msg or "%r != %r" % (a, b)
AssertionError: b'SELECT FOO FROM I_DONT_EXIST' != 'SELECT FOO FROM I_DONT_EXIST'
________________ UnicodeTest_mysql_mysqldb.test_native_unicode _________________
Traceback (most recent call last):
  File "$path/test/sql/test_types.py", line 943, in test_native_unicode
    expected
  File "$path/test/../lib/sqlalchemy/testing/assertions.py", line 187, in eq_
    assert a == b, msg or "%r != %r" % (a, b)
AssertionError: True != False
=========================== short test summary info ============================
FAIL test/dialect/mysql/test_types.py::TypesTest_mysql_mysqldb::()::test_bit_50_roundtrip
FAIL test/engine/test_execute.py::CompiledCacheTest_mysql_mysqldb::()::test_cache
FAIL test/engine/test_execute.py::CompiledCacheTest_mysql_mysqldb::()::test_keys_independent_of_ordering
FAIL test/engine/test_execute.py::EngineEventsTest_mysql_mysqldb::()::test_cursor_events_ctx_execute_scalar
FAIL test/engine/test_execute.py::EngineEventsTest_mysql_mysqldb::()::test_cursor_events_execute
FAIL test/engine/test_execute.py::EngineEventsTest_mysql_mysqldb::()::test_execute_events
FAIL test/engine/test_execute.py::ExecuteTest_mysql_mysqldb::()::test_unicode_test_fails_warning
FAIL test/engine/test_execute.py::HandleErrorTest_mysql_mysqldb::()::test_exception_event_non_dbapi_error
FAIL test/engine/test_execute.py::HandleErrorTest_mysql_mysqldb::()::test_handle_error
FAIL test/engine/test_execute.py::HandleErrorTest_mysql_mysqldb::()::test_legacy_dbapi_error
FAIL test/sql/test_types.py::UnicodeTest_mysql_mysqldb::()::test_native_unicode
======= 11 failed, 1255 passed, 424 skipped, 30 error in 158.28 seconds ========
```

--- a/_mysql.c
+++ b/_mysql.c
@@ -1372,7 +1372,8 @@ _mysql_field_to_python(
 		field_type == FIELD_TYPE_LONG_BLOB ||
 		field_type == FIELD_TYPE_BLOB ||
 		field_type == FIELD_TYPE_VAR_STRING ||
-		field_type == FIELD_TYPE_STRING) {
+		field_type == FIELD_TYPE_STRING ||
+		field_type == FIELD_TYPE_BIT) {
 			binary = 1;
 	}
 #endif

--- a/tests/test_MySQLdb_capabilities.py
+++ b/tests/test_MySQLdb_capabilities.py
@@ -1,10 +1,13 @@
 #!/usr/bin/env python
 import capabilities
+from datetime import timedelta
 import unittest
 import MySQLdb
 import warnings
 
+
 warnings.filterwarnings('ignore')
+
 
 class test_MySQLdb(capabilities.DatabaseTest):
 
@@ -13,12 +16,11 @@ class test_MySQLdb(capabilities.DatabaseTest):
     connect_kwargs = dict(use_unicode=True, sql_mode="ANSI,STRICT_TRANS_TABLES,TRADITIONAL")
     create_table_extra = "ENGINE=INNODB CHARACTER SET UTF8"
     leak_test = False
-    
+
     def quote_identifier(self, ident):
         return "`%s`" % ident
 
     def test_TIME(self):
-        from datetime import timedelta
         def generator(row,col):
             return timedelta(0, row*8000)
         self.check_data_integrity(
@@ -27,7 +29,7 @@ class test_MySQLdb(capabilities.DatabaseTest):
 
     def test_TINYINT(self):
         # Number data
-        def generator(row,col):
+        def generator(row, col):
             v = (row*row) % 256
             if v > 127:
                 v = v-256
@@ -35,7 +37,7 @@ class test_MySQLdb(capabilities.DatabaseTest):
         self.check_data_integrity(
             ('col1 TINYINT',),
             generator)
-        
+
     def test_stored_procedures(self):
         db = self.connection
         c = self.cursor
@@ -43,7 +45,7 @@ class test_MySQLdb(capabilities.DatabaseTest):
         c.executemany("INSERT INTO %s (pos,tree) VALUES (%%s,%%s)" % self.table,
                       list(enumerate('ash birch cedar larch pine'.split())))
         db.commit()
-        
+
         c.execute("""
         CREATE PROCEDURE test_sp(IN t VARCHAR(255))
         BEGIN
@@ -57,7 +59,7 @@ class test_MySQLdb(capabilities.DatabaseTest):
         self.assertEquals(len(rows), 1)
         self.assertEquals(rows[0][0], 3)
         c.nextset()
-        
+
         c.execute("DROP PROCEDURE test_sp")
         c.execute('drop table %s' % (self.table))
 
@@ -71,7 +73,29 @@ class test_MySQLdb(capabilities.DatabaseTest):
         self.check_data_integrity(
             ('col1 char(1)','col2 char(1)'),
             generator)
-    
+
+    def test_BIT(self):
+        c = self.cursor
+        try:
+            c.execute("""create table test_BIT (
+                b3 BIT(3),
+                b7 BIT(10),
+                b64 BIT(64))""")
+
+            one64 = '1'*64
+            c.execute(
+                "insert into test_BIT (b3, b7, b64)"
+                " VALUES (b'011', b'1111111111', b'%s')"
+                % one64)
+
+            c.execute("SELECT b3, b7, b64 FROM test_BIT")
+            row = c.fetchone()
+            self.assertEqual(row[0], b'\x03')
+            self.assertEqual(row[1], b'\x03\xff')
+            self.assertEqual(row[2], 'b\xff'*8)
+        finally:
+            c.execute("drop table if exists test_BIT")
+
     def test_bug_2671682(self):
         from MySQLdb.constants import ER
         try:
@@ -102,7 +126,7 @@ class test_MySQLdb(capabilities.DatabaseTest):
             return
         self.fail("Should raise ProgrammingError")
 
-        
+
 if __name__ == '__main__':
     if test_MySQLdb.leak_test:
         import gc

--- a/tests/test_MySQLdb_capabilities.py
+++ b/tests/test_MySQLdb_capabilities.py
@@ -92,7 +92,7 @@ class test_MySQLdb(capabilities.DatabaseTest):
             row = c.fetchone()
             self.assertEqual(row[0], b'\x03')
             self.assertEqual(row[1], b'\x03\xff')
-            self.assertEqual(row[2], 'b\xff'*8)
+            self.assertEqual(row[2], b'\xff'*8)
         finally:
             c.execute("drop table if exists test_BIT")
 

